### PR TITLE
OpenSSL 3.0: TLS MAC handling

### PR DESCRIPTION
**Pull request by fredmorcos**
### Short description
OpenSSL 3.0: TLS MAC handling. Use `EVP_MAC_CTX` instead of `HMAC_CTX` on OpenSSL 3.0 but continue to support `HMAC_CTX` for OpenSSL < 3.0.

### Checklist
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by fredmorcos** [comment]
@rgacogne Please double-check the build guards.

File: pdns/libssl.cc [hunk 1]
```diff
@@ -707,18 +721,74 @@ int OpenSSLTLSTicketKey::encrypt(unsigned char keyName[TLS_TICKETS_KEY_NAME_SIZE
...
 
+#if OPENSSL_VERSION_MAJOR >= 3
+  using ParamsBuilder = std::unique_ptr<OSSL_PARAM_BLD, decltype(&OSSL_PARAM_BLD_free)>;
+
+  auto params_build = ParamsBuilder(OSSL_PARAM_BLD_new(), OSSL_PARAM_BLD_free);
+  if (params_build == nullptr) {
+    return -1;
+  }
+
+  if (OSSL_PARAM_BLD_push_utf8_string(params_build.get(), OSSL_MAC_PARAM_DIGEST, "sha256", 7) == 0) {
```
Comment (rgacogne) [inline pdns/libssl.cc:732]: It might be nice to move "sha256" to a constant, but more importantly, what is `7`?

Comment (omoerbeek) [reply pdns/libssl.cc:732]: "The length of the string bsize should not include the terminating NUL byte." So it dos not look like the length of "sha256"

Comment (omoerbeek) [reply pdns/libssl.cc:732]: You can also use 0 to let the function compute the length.

Comment (fredmorcos) [reply pdns/libssl.cc:732]: I used 7 as the length of "sha256" and missed the part about not counting the terminating NUL byte. I also didn't want the function the re-compute something we already know the length of.

Does the following look OK?

```
constexpr static const char* sha256KeyName{"sha256"};
constexpr static const size_t sha256KeyNameLength = 6;
```

Comment (rgacogne) [reply pdns/libssl.cc]: Understood, I thought I had missed something! :-) Yes, the constants look good, thank you!

**Comment by fredmorcos** [comment]
Reminder to force-merge this please.

File: pdns/libssl.hh [hunk 2]
```diff
@@ -91,8 +91,17 @@ public:
   ~OpenSSLTLSTicketKey();
 
   bool nameMatches(const unsigned char name[TLS_TICKETS_KEY_NAME_SIZE]) const;
-  int encrypt(unsigned char keyName[TLS_TICKETS_KEY_NAME_SIZE], unsigned char *iv, EVP_CIPHER_CTX *ectx, HMAC_CTX *hctx) const;
-  bool decrypt(const unsigned char* iv, EVP_CIPHER_CTX *ectx, HMAC_CTX *hctx) const;
+
+#if OPENSSL_VERSION_MAJOR >= 3
+  constexpr static const char* sha256KeyName{"sha256"};
+  constexpr static const size_t sha256KeyNameLength{6};
```
Comment (rgacogne) [inline pdns/libssl.hh:97]: Would you mind moving these to `libssl.cc` instead?

**Review by rgacogne — CHANGES_REQUESTED** [review]
This currently does not build for me because we need to update the `ticket_key_callback` function and co in `doh.cc` as well.

File: pdns/libssl.hh (hunk 2 above)
Comment (cmouse) [reply pdns/libssl.hh:97]: why not just use std::string? you can then use c_str()

Comment (omoerbeek) [reply pdns/libssl.hh:97]: > why not just use std::string? you can then use c_str()

and `.size()`.

Comment (fredmorcos) [reply pdns/libssl.hh:97]: I don't think it's possible to use `std::string` with `constexpr`?
I suspect the compiler's constant folding mechanism should kick in, but I'm never sure in C++ so I figured it's just better to be explicit about what I want.

Comment (omoerbeek) [reply pdns/libssl.hh:97]: `constexpr` strings are in C++20 I believe. But I prefer `const` over `constexpr` if it allows to enhance readability.

**Comment by fredmorcos** [comment]
> This currently does not build for me because we need to update the `ticket_key_callback` function and co in `doh.cc` as well.

Fixed.

**Review by rgacogne — DISMISSED** [review]
:+1:

**Review by rgacogne — CHANGES_REQUESTED** [review]
We need that change as well:
```diff
diff --git pdns/dnsdistdist/doh.cc pdns/dnsdistdist/doh.cc
index ed2a165cf..443b15008 100644
--- pdns/dnsdistdist/doh.cc
+++ pdns/dnsdistdist/doh.cc
@@ -1390,7 +1390,11 @@ static void setupTLSContext(DOHAcceptContext& acceptCtx,

   if (tlsConfig.d_enableTickets && tlsConfig.d_numberOfTicketsKeys > 0) {
     acceptCtx.d_ticketKeys = std::make_unique<OpenSSLTLSTicketKeysRing>(tlsConfig.d_numberOfTicketsKeys);
+#if OPENSSL_VERSION_MAJOR >= 3
+    SSL_CTX_set_tlsext_ticket_key_evp_cb(ctx.get(), &ticket_key_callback);
+#else
     SSL_CTX_set_tlsext_ticket_key_cb(ctx.get(), &ticket_key_callback);
+#endif
     libssl_set_ticket_key_callback_data(ctx.get(), &acceptCtx);
   }
```

**Comment by fredmorcos** [comment]
Hopefully done now :-D

**Review by rgacogne — APPROVED** [review]

**Pull request merged**